diff --git a/project_browser.routing.yml b/project_browser.routing.yml
index b7ad176..9056bf8 100644
--- a/project_browser.routing.yml
+++ b/project_browser.routing.yml
@@ -15,7 +15,7 @@ project_browser.api_project_get_all:
     _controller: '\Drupal\project_browser\Controller\ProjectBrowserEndpointController::getAllProjects'
     _title: 'Get all projects'
   requirements:
-    _permission: 'administer modules'
+    _access: 'TRUE'
   #options:
   #  no_cache: 'TRUE'
 project_browser.browse:
diff --git a/sveltejs/public/build/bundle.js b/sveltejs/public/build/bundle.js
index 77726f3..638c305 100644
Binary files a/sveltejs/public/build/bundle.js and b/sveltejs/public/build/bundle.js differ
diff --git a/sveltejs/public/build/bundle.js.map b/sveltejs/public/build/bundle.js.map
index b8b9abd..e1d841c 100644
Binary files a/sveltejs/public/build/bundle.js.map and b/sveltejs/public/build/bundle.js.map differ
diff --git a/sveltejs/src/ProjectBrowser.svelte b/sveltejs/src/ProjectBrowser.svelte
index 8bf58aa..0ea1d82 100644
--- a/sveltejs/src/ProjectBrowser.svelte
+++ b/sveltejs/src/ProjectBrowser.svelte
@@ -102,9 +102,9 @@
       );
     }
 
-    const url = Drupal.url(
-      `drupal-org-proxy/project?${searchParams.toString()}`,
-    );
+    const url = `${window.location.protocol}//${
+      window.location.host
+    }/drupal-org-proxy/project?${searchParams.toString()}`;
 
     const res = await fetch(url);
     if (res.ok) {
