diff --git a/core/modules/block/block.module b/core/modules/block/block.module
index cb07d8ae43..c8304aab01 100644
--- a/core/modules/block/block.module
+++ b/core/modules/block/block.module
@@ -118,7 +118,7 @@ function block_themes_installed($theme_list) {
  * @param $theme
  *   The name of a theme.
  */
-function block_theme_initialize($theme) {
+function block_theme_initialize($theme) { return;
   // Initialize theme's blocks if none already registered.
   $has_blocks = \Drupal::entityTypeManager()->getStorage('block')->loadByProperties(['theme' => $theme]);
   if (!$has_blocks) {
